llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -167,20 +216,143 @@ std::pair<StringRef, unsigned> getSpecConstMetadata(Instruction *I) {
+    Desc.Offset = Offset;
+    Desc.Size = Ty->getPrimitiveSizeInBits() / 8;
+    Result[Index++] = Desc;
+    Offset += Desc.Size;
+  }
+}
+
+std::pair<StringRef, std::vector<CompositeSpecConstDescriptor>>
+getCompositeSpecConstMetadata(const Instruction *I) {
+  const MDNode *N = I->getMetadata(SPEC_CONST_SYM_ID_MD_STRING);
+  if (!N)
+    return std::make_pair("", std::vector<CompositeSpecConstDescriptor>{});
+  const auto *MDSym = cast<MDString>(N->getOperand(0));
+
+  std::vector<CompositeSpecConstDescriptor> Result(N->getNumOperands() - 1);
+  unsigned Index = 0, Offset = 0;
+  collectCompositeElementsInfoRecursive(I->getType(), Index, Offset, Result);
+
+  for (unsigned I = 1; I < N->getNumOperands(); ++I) {
+    const auto *MDInt = cast<ConstantAsMetadata>(N->getOperand(I));
+    unsigned ID = static_cast<unsigned>(
+        cast<ConstantInt>(MDInt->getValue())->getValue().getZExtValue());
+    Result[I - 1].ID = ID;
+  }
+  return std::make_pair(MDSym->getString(), Result);
+}
+
+Instruction *emitSpecConstant(unsigned NumericID, Type *Ty,
+                              Instruction *InsertBefore) {
+  Function *F = InsertBefore->getFunction();
+  // Generate arguments needed by the SPIRV version of the intrinsic
+  // - integer constant ID:
+  Value *ID = ConstantInt::get(Type::getInt32Ty(F->getContext()), NumericID);
+  // - default value:
+  Value *Def = getDefaultCPPValue(Ty);
+  // ... Now replace the call with SPIRV intrinsic version.
+  Value *Args[] = {ID, Def};
+  constexpr size_t NArgs = sizeof(Args) / sizeof(Args[0]);
+  Type *ArgTys[NArgs] = {nullptr};
+  for (unsigned int I = 0; I < NArgs; ++I)
+    ArgTys[I] = Args[I]->getType();
+  FunctionType *FT = FunctionType::get(Ty, ArgTys, false /*isVarArg*/);
+  Module *M = F->getParent();
+  std::string SPIRVName = mangleFuncItanium(SPIRV_GET_SPEC_CONST_VAL, FT);
+  FunctionCallee FC = M->getOrInsertFunction(SPIRVName, FT);
+  assert(FC.getCallee() && "SPIRV intrinsic creation failed");
+  CallInst *SpecConstant =
+      CallInst::Create(FT, FC.getCallee(), Args, "", InsertBefore);
+  return SpecConstant;
+}
+
+Instruction *emitSpecConstantComposite(Type *Ty,
+                                       ArrayRef<Instruction *> Elements,
+                                       Instruction *InsertBefore) {
+  SmallVector<Type *, 8> ArgTys(Elements.size());
+  SmallVector<Value *, 8> Args(Elements.size());
+  for (unsigned I = 0; I < Elements.size(); ++I) {
+    ArgTys[I] = Elements[I]->getType();
+    Args[I] = cast<Value>(Elements[I]);
+  }
+  auto *FT = FunctionType::get(Ty, ArgTys, false /*isVarArg*/);
+  Module *M = InsertBefore->getFunction()->getParent();
+  std::string SPIRVName = mangleFuncItanium(SPIRV_GET_SPEC_CONST_COMPOSITE, FT);
+  FunctionCallee FC = M->getOrInsertFunction(SPIRVName, FT);
+  assert(FC.getCallee() && "SPIRV intrinsic creation failed");
+  CallInst *SpecConstant =
+      CallInst::Create(FT, FC.getCallee(), Args, "", InsertBefore);
+  return SpecConstant;
+}
```

Comment:
These functions do similar things, for example they both create a function, add it to a module and create a call instruction. Common things can be moved to a separate function.

Reply:
Refactored in 4c9ce329cecc0563d51fd71aa56696fa2ff40dba